clients/src/test/java/org/apache/kafka/clients/NetworkClientTest.java

```diff
@@ -200,23 +202,112 @@ public void testRequestTimeout() {
         assertTrue("Expected response to fail due to disconnection", clientResponse.wasDisconnected());
     }
 
+    @Test
+    public void testConnectionThrottling() {
+        // Instrument the test to return a response with a 100ms throttle delay.
+        awaitReady(client, node);
+        ProduceRequest.Builder builder = ProduceRequest.Builder.forCurrentMagic((short) 1, 1000,
+            Collections.<TopicPartition, MemoryRecords>emptyMap());
+        TestCallbackHandler handler = new TestCallbackHandler();
+        ClientRequest request = client.newClientRequest(node.idString(), builder, time.milliseconds(), true, handler);
+        client.send(request, time.milliseconds());
+        client.poll(1, time.milliseconds());
+        ResponseHeader respHeader = new ResponseHeader(request.correlationId());
+        Struct resp = new Struct(ApiKeys.PRODUCE.responseSchema(ApiKeys.PRODUCE.latestVersion()));
+        resp.set("responses", new Object[0]);
+        resp.set(CommonFields.THROTTLE_TIME_MS, 100);
+        Struct responseHeaderStruct = respHeader.toStruct();
+        int size = responseHeaderStruct.sizeOf() + resp.sizeOf();
+        ByteBuffer buffer = ByteBuffer.allocate(size);
+        responseHeaderStruct.writeTo(buffer);
+        resp.writeTo(buffer);
+        buffer.flip();
+        selector.completeReceive(new NetworkReceive(node.idString(), buffer));
+        List<ClientResponse> responses = client.poll(1, time.milliseconds());
+
+        // The connection is not ready due to throttling.
+        assertFalse(client.ready(node, time.milliseconds()));
+        assertEquals(100, client.throttleDelayMs(node, time.milliseconds()));
+
+        // After 50ms, the connection is not ready yet.
+        time.sleep(50);
+        assertFalse(client.ready(node, time.milliseconds()));
+        assertEquals(50, client.throttleDelayMs(node, time.milliseconds()));
+
+        // After another 50ms, the throttling is done and the connection becomes ready again.
+        time.sleep(50);
+        assertTrue(client.ready(node, time.milliseconds()));
+        assertEquals(0, client.throttleDelayMs(node, time.milliseconds()));
+    }
+
+    // Sets expected ApiVersionsResponse from the specified node, where the max protocol version for API_VERSIONS is set
+    // to the specified version.
+    private void setExpectedApiVersionsResponse(Node node, short apiVersionsMaxProtocolVersion) {
```

Comment:
It may be slightly better to share the code with the existing `setExpectedApiVersionsResponse()`.

Reply:
Refactored the code to maximize code reuse.